src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -0,0 +1,139 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import jenkins.slaves.WorkspaceLocator;
+import org.apache.commons.codec.binary.Base32;
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Chooses manageable workspace names for branch projects.
+ * @see "JENKINS-34564"
+ */
+@Restricted(NoExternalUse.class)
+@Extension(ordinal = -100)
+public class WorkspaceLocatorImpl extends WorkspaceLocator {
+
+    private static final Logger LOGGER = Logger.getLogger(WorkspaceLocatorImpl.class.getName());
+
+    /** The most characters to allow in a workspace directory name, relative to the root. Zero to disable altogether. */
+    // TODO 2.4+ use SystemProperties
+    @SuppressWarnings("FieldMayBeFinal")
+    private static /* not final */ int PATH_MAX = Integer.getInteger(WorkspaceLocatorImpl.class.getName() + ".PATH_MAX", 80);
+
+    @Override
+    public FilePath locate(TopLevelItem item, Node node) {
+        if (PATH_MAX == 0) {
+            return null;
+        }
+        if (!(item.getParent() instanceof MultiBranchProject)) {
+            return null;
+        }
+        String minimized = minimize(item.getFullName());
+        if (node instanceof Jenkins) {
+            return ((Jenkins) node).getRootPath().child("workspace/" + minimized);
+        } else if (node instanceof Slave) {
+            FilePath root = ((Slave) node).getWorkspaceRoot();
+            return root != null ? root.child(minimized) : null;
+        } else { // ?
+            return null;
+        }
+    }
+
+    static String uniqueSuffix(String name) {
+        // TODO still in beta: byte[] sha256 = Hashing.sha256().hashString(name).asBytes();
+        byte[] sha256;
+        try {
+            sha256 = MessageDigest.getInstance("SHA-256").digest(name.getBytes(Charsets.UTF_16LE));
+        } catch (NoSuchAlgorithmException x) {
+            throw new AssertionError("https://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#MessageDigest", x);
+        }
+        return new Base32(0).encodeToString(sha256).replaceFirst("=+$", "");
+    }
+
+    static String minimize(String name) {
+        String mnemonic = name.replaceAll("(%[0-9A-F]{2}|[^a-zA-Z0-9-_.])+", "_");
+        int maxMnemonic = Math.max(PATH_MAX - /* ceil(256 / lg(32)) + length("-") */53, 1);
+        String result = StringUtils.right(mnemonic, maxMnemonic) + "-" + uniqueSuffix(name);
+        assert result.length() <= PATH_MAX : result + " does not fit inside " + PATH_MAX;
+        return result;
+    }
+
+    /**
+     * Cleans up workspace when an orphaned branch project is deleted.
+     * @see "JENKINS-2111"
+     */
+    @Extension
+    public static class Deleter extends ItemListener {
+
+        @Override
+        public void onDeleted(Item item) {
+            if (item.getParent() instanceof MultiBranchProject) {
```

Comment:
Could perhaps skip this block if `PATH_MAX == 0`, to save a bit of time; though it is still potentially useful in case the feature was active in a previous session and workspaces were created.